File: backend/app/schemas/performance.py
from typing import Annotated, Literal

from pydantic import Field

from .contracts import APIModel

UnitInterval = Annotated[float, Field(ge=0, le=1)]


class ConfusionMatrix(APIModel):
    true_positive: int
    false_positive: int
    true_negative: int
    false_negative: int


class ClassificationMetrics(APIModel):
    accuracy: UnitInterval
    precision: UnitInterval
    recall: UnitInterval
    f1: UnitInterval
    roc_auc: UnitInterval | None
    brier_score: UnitInterval
    log_loss: float
    threshold: float = 0.5
    confusion_matrix: ConfusionMatrix


class RegressionMetrics(APIModel):
    mae: float
    rmse: float
    r2: float


class EvaluationMetrics(APIModel):
    records: int
    classification: ClassificationMetrics
    regression: RegressionMetrics


class CandidateEvaluation(APIModel):
    algorithm: str
    validation: EvaluationMetrics
    test: EvaluationMetrics
    validation_selection_loss: float


class DisplayMetric(APIModel):
    name: str
    value: str
    description: str


class FeatureImportance(APIModel):
    name: str
    importance: float


class SplitInformation(APIModel):
    seed: int
    train: int
    validation: int
    test: int
    strategy: str = "60/20/20 stratified random split on the binary target; preprocessing fitted on training only."


class ModelPerformance(APIModel):
    version: str
    last_trained: str
    training_records: int
    features: int
    metrics: list[DisplayMetric]
    confusion_matrix: ConfusionMatrix
    feature_importance: list[FeatureImportance]
    is_demo: Literal[True] = True
    source: Literal["Synthetic"] = "Synthetic"
    algorithm: str
    feature_schema_version: str
    transformed_features: int
    feature_importance_method: str = "Classifier global tree impurity/gain importance (%), not per-project explanation or SHAP; may be biased."
    notice: str
    target: str
    split: SplitInformation
    dataset_version: str
    dataset_sha256: str
    selection_rule: str
    evaluations: list[CandidateEvaluation]
    baseline: CandidateEvaluation
    evaluation_scope: Literal["synthetic-held-out-test"] = "synthetic-held-out-test"
    probability_notice: str = "Class probabilities are uncalibrated. Brier score and log loss are reported, not a guarantee of calibration."

File: lib/landsight/performance.ts
import { MLAPIError, requestML } from './ml'
import type { ModelPerformance } from './types'

export function validateModelPerformance(value: unknown): ModelPerformance {
  const model = value as ModelPerformance | null
  const count = (n: unknown) => typeof n === 'number' && Number.isInteger(n) && n >= 0
  if (!model || model.evaluationScope !== 'synthetic-held-out-test' || model.isDemo !== true
    || !['version', 'lastTrained', 'algorithm', 'notice', 'featureImportanceMethod', 'featureSchemaVersion', 'target', 'probabilityNotice'].every(key => typeof model[key as keyof ModelPerformance] === 'string')
    || !Number.isFinite(Date.parse(model.lastTrained)) || !count(model.trainingRecords) || !count(model.features)
    || !Array.isArray(model.metrics) || !model.metrics.length || model.metrics.some(m => !m || typeof m.name !== 'string' || typeof m.value !== 'string' || typeof m.description !== 'string')
    || !model.confusionMatrix || !['truePositive', 'falsePositive', 'trueNegative', 'falseNegative'].every(k => count(model.confusionMatrix[k as keyof typeof model.confusionMatrix]))
    || !Array.isArray(model.featureImportance) || !model.featureImportance.length || model.featureImportance.some(f => !f || typeof f.name !== 'string' || !Number.isFinite(f.importance) || f.importance < 0 || f.importance > 100)
    || !model.split || ![model.split.train, model.split.validation, model.split.test].every(count) || typeof model.split.strategy !== 'string'
    || model.trainingRecords !== model.split.train
    || Object.values(model.confusionMatrix).reduce((sum, n) => sum + n, 0) !== model.split.test) {
    throw new MLAPIError('invalid-response', 'Invalid measured model-performance response.')
  }
  return model
}

export async function fetchModelPerformance() {
  return validateModelPerformance(await requestML('/api/ml/model-performance'))
}
